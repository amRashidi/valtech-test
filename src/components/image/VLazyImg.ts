/* eslint-disable indent */
import {
    reactive,
    h,
    computed,
    ref,
    onMounted,
    onBeforeUnmount,
    defineComponent,
    Ref,
} from "vue";
import noImage from "/error.jpg";
import loadingImage from "/loading.png";
interface IObserverState {
    observer: IntersectionObserver | null;
    intersected: boolean;
    error: boolean;
    loaded: boolean;
}
export default defineComponent({
    props: {
        src: {
            type: String,
            default: "",
            required: true,
        },
        srcError: {
            type: String,
            default: noImage,
        },
        srcPlaceholder: {
            type: String,
            default: noImage,
        },
        srcset: {
            type: String,
        },
        intersectionOptions: {
            type: Object,
            default: () => ({}),
        },
        usePicture: {
            type: Boolean,
            default: false,
        },
    },
    inheritAttrs: false,
    setup(props, { attrs, slots, emit }) {
        const root: Ref<Element | null> = ref(null);
        const state: IObserverState = reactive({
            observer: null,
            intersected: false,
            error: false,
            loaded: false,
        });

        // Computed
        const srcImage = computed(() => {
            let src = loadingImage;
            if (state.intersected && props.src) src = props.src;
            if (state.error) {
                src = props.srcError;
            }
            return src;
        });
        const srcsetImage = computed(() =>
            state.intersected && props.srcset ? props.srcset : false
        );
        const load = () => {
            if (
                root.value &&
                root.value.getAttribute("src") !== props.srcPlaceholder
            ) {
                state.loaded = true;
                emit("load", root.value);
            }
        };
        const error = () => {
            state.error = true;
            emit("error", root.value);
        };
        onMounted(() => {
            if ("IntersectionObserver" in window) {
                state.observer = new IntersectionObserver((entries) => {
                    const image = entries[0];
                    if (image.isIntersecting) {
                        state.intersected = true;
                        state.observer && state.observer.disconnect();
                        emit("intersect");
                    }
                }, props.intersectionOptions);

                state.observer.observe(root.value as Element);
            }
        });
        onBeforeUnmount(() => {
            if ("IntersectionObserver" in window && state.observer) {
                state.observer.disconnect();
            }
        });

        return () => {
            const img = h("img", {
                ref: root,
                src: srcImage.value,
                srcset: srcsetImage.value || null,
                ...attrs,
                class: [
                    attrs.class,
                    "v-lazy-image",
                    { "v-lazy-image-loaded": state.loaded },
                ],
                onLoad: load,
                onError: error,
            });
            return img;
        };
    },
});