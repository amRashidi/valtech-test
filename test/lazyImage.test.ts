import { flushPromises, mount } from '@vue/test-utils';
import VLazyImg from "../src/components/image/VLazyImg";

test('mount img', () => {
    expect(VLazyImg).toBeTruthy()
    const img = mount(VLazyImg)
    expect(img.classes()).toContain("v-lazy-image");
})
test('attr img', () => {
    expect(VLazyImg).toBeTruthy()
    const img = mount(VLazyImg, {
        props: {
            src: "/cities/venice.jpg",
        }
    })
    expect(img.html()).toMatchSnapshot()
})
test('img loading src', async () => {
    expect(VLazyImg).toBeTruthy()
    const img = mount(VLazyImg)
    await img.setProps({ src: "/cities/venice.jpg" })
    expect(img.attributes('src')).toEqual("/loading.png")
})