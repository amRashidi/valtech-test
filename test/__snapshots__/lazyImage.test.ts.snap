// Vitest Snapshot v1

exports[`attr img 1`] = `"<img src=\\"/loading.png\\" class=\\"v-lazy-image\\">"`;
